# Sprint 10 - Tasca M10 T01


## Exercici 1

Realitza web scraping de dues de les tres pàgines web proposades utilitzant BeautifulSoup primer i Selenium després. 

- http://quotes.toscrape.com

- https://www.bolsamadrid.es

- www.wikipedia.es (fes alguna cerca primer i escrapeja algun contingut)



## Exercici 2

Documenta en un Word el teu conjunt de dades generat amb la informació que tenen els diferents arxius de Kaggle.



Documentación y tutoriales:

* Beautiful Soup:
    - [Documentación](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
    - [Doc 2](https://beautiful-soup-4.readthedocs.io/en/latest/#)
    - [Tutorial](https://realpython.com/beautiful-soup-web-scraper-python/)
    - [Tutorial 2](https://towardsdatascience.com/step-by-step-tutorial-web-scraping-wikipedia-with-beautifulsoup-48d7f2dfa52d)
    - [Link](https://www.datacamp.com/tutorial/web-scraping-python-nlp)
    - [Link 2](https://www.datacamp.com/tutorial/web-scraping-using-python)
    


* Selenium:
    - [Documentación](https://www.selenium.dev/documentation/webdriver/getting_started/)
    - [Automatización](https://realpython.com/modern-web-automation-with-python-and-selenium/)
    - [Tutorial avanzado](https://selenium-python.readthedocs.io/getting-started.html#simple-usage)
    
    
 * Otros:
     - [Avoid Blocked](https://www.codementor.io/@scrapingdog/10-tips-to-avoid-getting-blocked-while-scraping-websites-16papipe62)
     - [Scrapy](https://scrapy.org/)
     - [Paso a paso](https://vegibit.com/python-web-scraping-with-beautiful-soup/)

# Ejercicio 1: 

In [1]:
from bs4 import BeautifulSoup as bs
import warnings
warnings.filterwarnings('ignore')
import requests
import pandas as pd

## Beautiful Soup:

#### a) http://quotes.toscrape.com

In [3]:
url_a = "http://quotes.toscrape.com"

In [4]:
r = requests.get(url_a)

In [5]:
r.status_code

200

In [6]:
soup = bs(r.content)

In [7]:
print(soup.prettify()) 

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   Quotes to Scrape
  </title>
  <link href="/static/bootstrap.min.css" rel="stylesheet"/>
  <link href="/static/main.css" rel="stylesheet"/>
 </head>
 <body>
  <div class="container">
   <div class="row header-box">
    <div class="col-md-8">
     <h1>
      <a href="/" style="text-decoration: none">
       Quotes to Scrape
      </a>
     </h1>
    </div>
    <div class="col-md-4">
     <p>
      <a href="/login">
       Login
      </a>
     </p>
    </div>
   </div>
   <div class="row">
    <div class="col-md-8">
     <div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
      <span class="text" itemprop="text">
       “The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
      </span>
      <span>
       by
       <small class="author" itemprop="author">
        Albert Einstein
       </small>
       <a href="/author/Albert

In [8]:
quote_elements = soup.find_all("div", class_="quote")
quote_elements

[<div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
 <span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>
 <span>by <small class="author" itemprop="author">Albert Einstein</small>
 <a href="/author/Albert-Einstein">(about)</a>
 </span>
 <div class="tags">
             Tags:
             <meta class="keywords" content="change,deep-thoughts,thinking,world" itemprop="keywords"/>
 <a class="tag" href="/tag/change/page/1/">change</a>
 <a class="tag" href="/tag/deep-thoughts/page/1/">deep-thoughts</a>
 <a class="tag" href="/tag/thinking/page/1/">thinking</a>
 <a class="tag" href="/tag/world/page/1/">world</a>
 </div>
 </div>,
 <div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
 <span class="text" itemprop="text">“It is our choices, Harry, that show what we truly are, far more than our abilities.”</span>
 <span>by <small class="author" itempr

In [10]:
df = pd.DataFrame()
for quote_element in quote_elements:
    quote = quote_element.find("span", class_="text")
    author = quote_element.find("small", class_="author")
    df = df.append({"Frase" : quote.text.strip(),
                            "Autor" : author.text.strip()},
                        ignore_index=True)

In [11]:
df

Frase              Autor
0  “The world as we have created it is a process ...    Albert Einstein
1  “It is our choices, Harry, that show what we t...       J.K. Rowling
2  “There are only two ways to live your life. On...    Albert Einstein
3  “The person, be it gentleman or lady, who has ...        Jane Austen
4  “Imperfection is beauty, madness is genius and...     Marilyn Monroe
5  “Try not to become a man of success. Rather be...    Albert Einstein
6  “It is better to be hated for what you are tha...         André Gide
7  “I have not failed. I've just found 10,000 way...   Thomas A. Edison
8  “A woman is like a tea bag; you never know how...  Eleanor Roosevelt
9  “A day without sunshine is like, you know, nig...       Steve Martin

Recogemos en un dataframe las frases y el autor de la primera página de la web a la que hemos realizado web scraping con Beautiful Soup. 

#### b) https://es.wikipedia.org/wiki/Esperanto

In [2]:
url_b = "https://es.wikipedia.org/wiki/Esperanto"

In [12]:
r_b = requests.get(url_b)

In [13]:
r_b.status_code

200

In [15]:
soup_b = bs(r_b.content)

In [16]:
print(soup_b.prettify()) 

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="es">
 <head>
  <meta charset="utf-8"/>
  <title>
   Esperanto - Wikipedia, la enciclopedia libre
  </title>
  <script>
   document.documentElement.className="client-js";(function(){var cookie=document.cookie.match(/(?:^|; )eswikimwclientprefs=([^;]+)/);if(cookie){var featureName=cookie[1];document.documentElement.className=document.documentElement.className.replace(featureName+'-enabled',featureName+'-disabled');}}());RLCONF={"wgBreakFrames":false,"wgSeparatorTransformTable":[",\t."," \t,"],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","enero","febrero","marzo","abril","mayo","junio","julio","agosto","septiembre","octubre","noviembre","diciembre"],"wgRequestId":"cc091da0-5f70-4d84-b386-0f40103a7b7b","wgCSPNonce":false,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":false,"wgNamespaceNumber":0,"wgPageName":"Esperanto","wgTitle":"Esperanto","wgCurRevisionId":149342303,"wgRevisionId":14934230

In [19]:
all_links = soup.find_all("a")
for link in all_links:
    print(link.get("href"))

None
/wiki/Wikipedia:Art%C3%ADculos_buenos
//es.wikipedia.org/wiki/Portal:Esperanto
#mw-head
#searchInput
/wiki/Esperanto_(desambiguaci%C3%B3n)
/wiki/Archivo:Flag_of_Esperanto.svg
/wiki/Archivo:Flag_of_Andorra.svg
/wiki/Andorra
/wiki/Archivo:Flag_of_Lithuania.svg
/wiki/Lituania
/wiki/Archivo:Flag_of_Iceland.svg
/wiki/Islandia
/wiki/Archivo:Flag_of_Hungary.svg
/wiki/Hungr%C3%ADa
/wiki/Archivo:Flag_of_Luxembourg.svg
/wiki/Luxemburgo
/wiki/Archivo:Flag_of_FLNKS.svg
/wiki/Nueva_Caledonia
/wiki/Archivo:Flag_of_Belgium_(civil).svg
/wiki/B%C3%A9lgica
/wiki/Archivo:Flag_of_Finland.svg
/wiki/Finlandia
/wiki/Archivo:Flag_of_France_(1794%E2%80%931815,_1830%E2%80%931974,_2020%E2%80%93present).svg
/wiki/Francia
/wiki/Archivo:Flag_of_Denmark.svg
/wiki/Dinamarca
/wiki/Archivo:Flag_of_Sweden.svg
/wiki/Suecia
/wiki/Archivo:Flag_of_Switzerland.svg
/wiki/Suiza
/wiki/Archivo:Flag_of_the_Netherlands.svg
/wiki/Pa%C3%ADses_Bajos
/wiki/Archivo:Flag_of_the_Czech_Republic.svg
/wiki/Rep%C3%BAblica_Checa
/wiki/Ar

In [24]:
rows = soup.find_all("a")
print(rows[:10])

[<a id="top"></a>, <a href="/wiki/Wikipedia:Art%C3%ADculos_buenos" title="Artículo bueno"><img alt="Artículo bueno" data-file-height="185" data-file-width="180" decoding="async" height="14" src="//upload.wikimedia.org/wikipedia/commons/thumb/e/e5/Art%C3%ADculo_bueno.svg/14px-Art%C3%ADculo_bueno.svg.png" srcset="//upload.wikimedia.org/wikipedia/commons/thumb/e/e5/Art%C3%ADculo_bueno.svg/21px-Art%C3%ADculo_bueno.svg.png 1.5x, //upload.wikimedia.org/wikipedia/commons/thumb/e/e5/Art%C3%ADculo_bueno.svg/28px-Art%C3%ADculo_bueno.svg.png 2x" width="14"/></a>, <a href="//es.wikipedia.org/wiki/Portal:Esperanto" title="El portal asociado a este artículo"><img alt="El portal asociado a este artículo" data-file-height="32" data-file-width="36" decoding="async" height="16" src="//upload.wikimedia.org/wikipedia/commons/thumb/c/c9/Portal.svg/18px-Portal.svg.png" srcset="//upload.wikimedia.org/wikipedia/commons/thumb/c/c9/Portal.svg/27px-Portal.svg.png 1.5x, //upload.wikimedia.org/wikipedia/commons/th

In [30]:
df_2 = pd.DataFrame(all_links)
df_2.head(30)

0
0                             []
1                           [[]]
2                           [[]]
3           [Ir a la navegación]
4             [Ir a la búsqueda]
5   [Esperanto (desambiguación)]
6                           [[]]
7                           [[]]
8                      [Andorra]
9                           [[]]
10                    [Lituania]
11                          [[]]
12                    [Islandia]
13                          [[]]
14                     [Hungría]
15                          [[]]
16                  [Luxemburgo]
17                          [[]]
18             [Nueva Caledonia]
19                          [[]]
20                     [Bélgica]
21                          [[]]
22                   [Finlandia]
23                          [[]]
24                     [Francia]
25                          [[]]
26                   [Dinamarca]
27                          [[]]
28                      [Suecia]
29                          [[]]

In [32]:
df_2.shape

(1771, 1)

En un dataframe recogemos todos los link, internos y externos que tiene la página seleccionada de wikipedia, esto resulta interesante por ejemplo de cara a ver técnicas de posicionamiento SEO, ya que en este caso es altisimo el número de link en un sólo artículo, alcanzando los 1771 links. 

#### b) https://es.wikipedia.org/wiki/Esperanto

## Selenium:

#### a) http://quotes.toscrape.com

In [33]:
df.to_csv('frases_celebres.csv', header=False, index=False)

In [ ]:
df_2.to_csv('frases_celebres.csv', header=False, index=False)